Mathlib/Analysis/NormedSpace/Connected.lean

```diff
@@ -0,0 +1,188 @@
+/-
+Copyright (c) 2023 Sébastien Gouëzel. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Sébastien Gouëzel
+-/
+import Mathlib.Analysis.Convex.Topology
+import Mathlib.LinearAlgebra.Dimension
+import Mathlib.Topology.Algebra.Module.Cardinality
+
+/-!
+# Connectedness of subsets of vector spaces
+
+We show several results related to the (path)-connectedness of subsets of real vector spaces:
+* `Set.Countable.isPathConnected_compl_of_one_lt_rank` asserts that the complement of a countable
+  set is path-connected in a space of dimension `> 1`.
+* `isPathConnected_compl_singleton_of_one_lt_rank` is the special case of the complement of a
+  singleton.
+* `isPathConnected_sphere` shows that any sphere is path-connected in dimension `> 1`.
+
+Statements with connectedness instead of path-connectedness are also given.
+-/
+
+open Convex Set Metric
+
+section TopologicalVectorSpace
+
+variable {E : Type*} [AddCommGroup E] [Module ℝ E]
+
+lemma segment_inter_eq_endpoint_of_linearIndependent
+    {x y : E} (h : LinearIndependent ℝ ![x, y]) {s t : ℝ} (hs : s ≠ t) (c : E) :
+    [c + x -[ℝ] c + t • y] ∩ [c + x -[ℝ] c + s • y] ⊆ {c + x} := by
+  intro z ⟨hzt, hzs⟩
+  rw [segment_eq_image, mem_image] at hzt hzs
+  rcases hzt with ⟨p, ⟨p0, p1⟩, rfl⟩
+  rcases hzs with ⟨q, ⟨q0, q1⟩, H⟩
+  have H' : (1 - q) • x + (q * s) • y = (1 - p) • x + (p * t) • y := by
+    rw [← sub_eq_zero] at H ⊢
+    rw [← H]
+    simp only [smul_smul, smul_add, sub_smul]
+    abel
+  obtain rfl : q = p := by simpa using (h.eq_of_pair H').1
+  rcases q0.eq_or_gt with rfl|hq0'
+  · simp
+  · have A : s = t := by simpa [mul_eq_mul_left_iff, hq0'.ne'] using (h.eq_of_pair H').2
+    exact (hs A).elim
+
+variable [TopologicalSpace E] [ContinuousAdd E] [ContinuousSMul ℝ E]
+
+/-- In a real vector space of dimension `> 1`, the complement of any countable set is path
+connected. -/
+theorem Set.Countable.isPathConnected_compl_of_one_lt_rank
+    (h : 1 < Module.rank ℝ E) {s : Set E} (hs : s.Countable) :
```

Comment:
Do you think there's a case for a typeclass to capture `h` here? I can't immediately think of a good name but maybe something like:
```lean
class ContainsPlane (K V : Type*) [Semiring K] [AddCommMonoid V] [Module K V] : Prop :=
  one_lt_rank : 1 < Module.rank K V
```
That way some of the results here could be instances / simp lemmas. More importantly, it would make the results here unopinionated about the way in which the concept "dimension > 1" is expressed which should, for example, provide a convenient means to make these results available in a situation where someone has `[FiniteDimensional K V]` and information about `finrank K V`.

Reply:
I think we should only implement this typeclass when there is something concrete to gain for it and when there are more lemmas that take advantage of it. (Also, what about higher-dimensional analogues?)

Reply:
Ah yes, the class would probably be something like:
```lean
class RankAtLeast (K V : Type*) [Semiring K] [AddCommMonoid V] [Module K V] (n : Cardinal) : Prop :=
  le_rank : n ≤ Module.rank K V
```
but I agree we can wait till the need is greater. (A quick scan of the library for the regex `(≤|<) (Module\.|fin|)rank` finds a few would-be instances but not really any consumers.)

Reply:
I encourage you to generalise + relocate `segment_inter_eq_endpoint_of_linearIndependent_sub` but I do not insist: I agree this is subjective.